oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/osh/OSHEntityTimeUtils.java

```diff
@@ -0,0 +1,357 @@
+   * child elements (useful to find out when the geometry of this object has been altered).
+   * </p>
+   *
+   * @param osh the osh entity to work on
+   * @param recurse specifies if times of modifications of child entities should also be returned or
+   *        not
+   * @return a list of timestamps where this entity has been modified
+   */
+  public static List<OSHDBTimestamp> getModificationTimestamps(OSHEntity osh, boolean recurse) {
+    switch (osh.getType()) {
+      case NODE:
+        return getModificationTimestamps((OSHNode) osh);
+      case WAY:
+      case RELATION:
+        return getModificationTimestamps(osh, recurse, null);
+      default:
+        throw new UnsupportedOperationException(format(UNSUPPORTED_OSMTYPE_MESSAGE, osh.getType()));
+    }
+  }
+
+  private static List<OSHDBTimestamp> getModificationTimestamps(OSHNode osh) {
+    List<OSHDBTimestamp> result = new ArrayList<>();
+    for (OSMEntity osm : osh.getVersions()) {
+      result.add(osm.getTimestamp());
+    }
+    return Lists.reverse(result);
+  }
+
+  /**
+   * Returns all timestamps at which this entity (or one or more of its child entities) has been
+   * modified and matches a given condition/filter.
+   *
+   * <p>
+   * Consecutive modifications from a single changeset are grouped together (only the last
+   * modification timestamp of the corresponding changeset is considered). This can reduce the
+   * amount of geometry modifications by a lot (e.g. when sequential node uploads of a way
+   * modification causes many intermediate modification states), making results more
+   * "accurate"/comparable as well as allowing faster processing of geometries.
+   * </p>
+   *
+   * @param osh the osh entity to work on
+   * @param osmEntityFilter only timestamps for which the entity matches this filter are returned
+   * @param changesetTimestamps association between timestamps and changeset-ids, can be obtained
+   *        from oshEntity by calling {@link #getChangesetTimestamps}.
+   * @return a list of timestamps where this entity has been modified
+   */
+  public static List<OSHDBTimestamp> getModificationTimestamps(OSHEntity osh,
+      Predicate<OSMEntity> osmEntityFilter, Map<OSHDBTimestamp, Long> changesetTimestamps) {
+    List<OSHDBTimestamp> allModificationTimestamps = getModificationTimestamps(osh, true, osmEntityFilter);
+
+    if (allModificationTimestamps.size() <= 1) {
+      return allModificationTimestamps;
+    }
+    // group modification timestamps by changeset
+    List<OSHDBTimestamp> result = new ArrayList<>(allModificationTimestamps.size());
+    allModificationTimestamps = Lists.reverse(allModificationTimestamps);
+    long nextChangeset = -1L;
+    for (OSHDBTimestamp timestamp : allModificationTimestamps) {
+      long changeset = changesetTimestamps.get(timestamp);
+      if (changeset != nextChangeset) {
+        result.add(timestamp);
+      }
+      nextChangeset = changeset;
+    }
+
+    return Lists.reverse(result);
```

Comment:
We could try to inplace group modification timestamps, to avoid the additional array creation and the reversing back and forth.
Here an untested suggestion:
```java
int pos = 0;
for (int i=0; i < allModificationTimestamps.size() ;) {
  var timestamp = allModificationTimestamps.get(i);
  var changeset = changesetTimestamps.get(timestamp);
  i++;
  for (; i < allModificationTimestamps.size(); i++) {
    var nextTimestamp = allModificationTimestamps.get(i);
    var nextChangeset = changesetTimestamps.get(nextTimestamp);
    if (changeset != nextChangeset) {
      break;
    }
    timestamp = nextTimestamp;
  }
  allModificationTimestamps.set(pos++, timestamp);
}
return allModificationTimestamps.subList(0, pos);
```

Reply:
good idea, but what do you think about this (IMHO) slightly less complex solution:

```java

    // group modification timestamps by changeset
    long nextChangeset = -1L;
    int pos = allModificationTimestamps.size();
    for (OSHDBTimestamp timestamp : Lists.reverse(allModificationTimestamps)) {
      long changeset = changesetTimestamps.get(timestamp);
      if (changeset != nextChangeset) {
        allModificationTimestamps.set(--pos, timestamp);
      }
      nextChangeset = changeset;
    }

    return allModificationTimestamps.subList(pos, allModificationTimestamps.size());
```

it comes at the cost of one additional `Lists.reverse`, but since that's just a lightweight view, I guess we ca live with it.